cardinal/ecs/cql/cql.go

```diff
@@ -133,3 +136,76 @@ func (t *cqlTerm) String() string {
 }
 
 var CQLParser = participle.MustBuild[cqlTerm]()
+
+// TODO: Value is sum type is represented as a product type. There is a case where multiple properties are filled out.
+// Only one property may not be nil, The parser should prevent this from happening but for safety this should eventually
+// be checked.
+func valueToLayoutFilter(value *cqlValue, stringToComponent func(string) component.IComponentType) (filter.LayoutFilter, error) {
+	if value.Not != nil {
+		result_filter, err := valueToLayoutFilter(value.Not.SubExpression, stringToComponent)
+		if err != nil {
+			return nil, err
+		}
+		return filter.Not(result_filter), nil
+	} else if value.Exact != nil {
+		if len(value.Exact.Components) <= 0 {
+			return nil, errors.New("EXACT cannot have zero parameters")
+		}
+		components := make([]component.IComponentType, 0, len(value.Exact.Components))
+		for _, componentName := range value.Exact.Components {
+			components = append(components, stringToComponent(componentName.Name))
+		}
+		return filter.Exact(components...), nil
+	} else if value.Contains != nil {
+		if len(value.Contains.Components) <= 0 {
+			return nil, errors.New("CONTAINS cannot have zero parameters")
+		}
+		components := make([]component.IComponentType, 0, len(value.Contains.Components))
+		for _, componentName := range value.Contains.Components {
+			components = append(components, stringToComponent(componentName.Name))
+		}
+		return filter.Contains(components...), nil
+	} else if value.Subexpression != nil {
+		return termToLayoutFilter(value.Subexpression, stringToComponent)
+	} else {
+		return nil, errors.New("unknown error during conversion from CQL AST to LayoutFilter")
+	}
+}
+
+func factorToLayoutFilter(factor *cqlFactor, stringToComponent func(string) component.IComponentType) (filter.LayoutFilter, error) {
+	return valueToLayoutFilter(factor.Base, stringToComponent)
+}
+
+func opFactorToLayoutFilter(opFactor *cqlOpFactor, stringToComponent func(string) component.IComponentType) (*cqlOperator, filter.LayoutFilter, error) {
+	resultFilter, err := factorToLayoutFilter(opFactor.Factor, stringToComponent)
+	if err != nil {
+		return nil, nil, err
+	}
+	return &opFactor.Operator, resultFilter, nil
+}
+
+func termToLayoutFilter(term *cqlTerm, stringToComponent func(string) component.IComponentType) (filter.LayoutFilter, error) {
+	if term.Left == nil {
+		return nil, errors.New("Not enough values in expression")
+	}
+	acc, err := factorToLayoutFilter(term.Left, stringToComponent)
+	if err != nil {
+		return nil, err
+	}
+	if len(term.Right) > 0 {
+		for _, opFactor := range term.Right {
+			operator, resultFilter, err := opFactorToLayoutFilter(opFactor, stringToComponent)
+			if err != nil {
+				return nil, err
+			}
+			if *operator == opAnd {
+				acc = filter.And(acc, resultFilter)
+			} else if *operator == opOr {
+				acc = filter.Or(acc, resultFilter)
+			} else {
+				return nil, errors.New("invalid operator")
+			}
```

Comment:
smol nit: this might look nicer as a switch case 

also, looks like we can just do return the filter instead of setting it to `acc` here

Reply:
that's fine, I'm used to python so i like if-else. 

I'm not sure what you mean by not returning the acc? The acc is all the filters combined?